[AArch64][SME2] Add ZT0 attributes to SMEAttrs

kmclaughlin-arm opened the pull request:
This patch extends SMEAttrs to interpret the following new attributes,
which are mutually exclusive and apply to SME2 only:
  - aarch64_sme_zt0_in (ZT0_In)
  - aarch64_sme_zt0_out (ZT0_Out)
  - aarch64_sme_zt0_inout (ZT0_InOut)
  - aarch64_sme_zt0_new (ZT0_New)
  - aarch64_sme_zt0_preserved (ZT0_Preserved)

ZT0_In, ZT0_Out, ZT0_InOut & ZT0_Preserved are all considered to share
ZT0. These attributes will be required by later patches to determine
if ZT0 should be preserved around function calls, or cleared on entry
to the function.

kmclaughlin-arm committed:
[AArch64][SME2] Add ZT0 attributes to SMEAttrs

This patch extends SMEAttrs to interpret the following new attributes,
which are mutually exclusive and apply to SME2 only:
  - aarch64_sme_zt0_in (ZT0_In)
  - aarch64_sme_zt0_out (ZT0_Out)
  - aarch64_sme_zt0_inout (ZT0_InOut)
  - aarch64_sme_zt0_new (ZT0_New)
  - aarch64_sme_zt0_preserved (ZT0_Preserved)

ZT0_In, ZT0_Out, ZT0_InOut & ZT0_Preserved are all considered to share
ZT0. These attributes will be required by later patches to determine
if ZT0 should be preserved around function calls, or cleared on entry
to the function.

MDevereau commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.cpp:
```diff
@@ -28,6 +31,11 @@ void SMEAttrs::set(unsigned M, bool Enable) {
          "ZA_Shared and ZA_NoLazySave are mutually exclusive");
+
+  // ZT0 Attrs
+  assert((!sharesZT0() || (hasNewZT0Body() ^ isZT0In() ^ isZT0InOut() ^
+                           isZT0Out() ^ preservesZT0())) &&
+         "ZT0_New,In,Out,InOut,Preserved are mutually exclusive");
```
Is it possible to tighten this up and make it more specific? First, I think it should have full names like `ZT0_New, ZT0_In, ZT0_Out` and have spaces after the commas to make it a bit easier to read. Second, sorry if I'm sending you on a wild goose chase, is it possible to do something along the lines of

```c++
SmallVector<StringRef> Attrs;
if (hasNewZT0Body())
  Attrs.push_back("ZT0_New");
if (isZT0In())
  Attrs.push_back("ZT0_In");
// ...
// test other mutually exclusive attributes

if (Attrs.size() > 1){
  // join Attrs into a string and bail as an llvm_unreachable
}
``` 
?
That way we print a better diagnostic since we know exactly what mutually exclusive attributes are causing the problem

sdesmalen-arm replied:
This is not really a diagnostic, but rather a message of an assert. Asserts aren't really user-facing, they are more a way to identify an issue, which has to be clear enough to understand what the issue is. From the message, we know that mutually exclusive attributes have been combined somehow, which is enough to start debugging and find the cause of it.

I agree with your comment about having spaces between the commas and having the full names.

kmclaughlin-arm replied:
I've reworded this assert to use the full names, it's now:
`ZT0_New, ZT0_In, ZT0_Out, ZT0_InOut and ZT0_Preserved are all mutually exclusive`

kmclaughlin-arm committed:
- Added a sharedZA() function which returns true if aarch64_pstate_za_shared is set
- Changed hasSharedZAInterface() to return true if either sharedZA() or sharedZT0() is true
- Changed wording of the assert checking for mutually exclusive ZT0 attributes

kmclaughlin-arm committed:
- Removed duplicate tests from SMEAttributesTest

kmclaughlin-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -50,6 +54,53 @@ TEST(SMEAttributes, Constructors) {
                       ->getFunction("foo"))
                   .preservesZA());
 
+  ASSERT_TRUE(SA(*parseIR("declare void @foo() \"aarch64_sme_zt0_in\"")
+                      ->getFunction("foo"))
+                  .isZT0In());
+  ASSERT_TRUE(SA(*parseIR("declare void @foo() \"aarch64_sme_zt0_out\"")
+                      ->getFunction("foo"))
+                  .isZT0Out());
+  ASSERT_TRUE(SA(*parseIR("declare void @foo() \"aarch64_sme_zt0_inout\"")
+                      ->getFunction("foo"))
+                  .isZT0InOut());
```
These are tested a bit further down, on line 93:

```
ASSERT_TRUE(SA(*parseIR("declare void @foo() \"aarch64_sme_zt0_preserved\"")
                    ->getFunction("foo"))
                .preservesZT0());

ASSERT_TRUE(SA(*parseIR("declare void @foo() \"aarch64_sme_zt0_new\"")
                    ->getFunction("foo"))
                .hasNewZT0Body());
```

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -111,36 +80,36 @@ TEST(SMEAttributes, Constructors) {
   EXPECT_DEBUG_DEATH(SA(SA::ZT0_New | SA::ZT0_In),
                      "ZT0_New, ZT0_In, ZT0_Out, ZT0_InOut and ZT0_Preserved "
-                     "are all mutually exclusive");
+                     "are all \" \"mutually exclusive");
```
Did something go wrong here with the escapes/formatting? :)

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -58,6 +78,39 @@ TEST(SMEAttributes, Constructors) {
+  EXPECT_DEBUG_DEATH(SA(SA::ZT0_New | SA::ZT0_In),
+                     "ZT0_New, ZT0_In, ZT0_Out, ZT0_InOut and ZT0_Preserved "
+                     "are all \" \"mutually exclusive");
+  EXPECT_DEBUG_DEATH(SA(SA::ZT0_New | SA::ZT0_Out),
```
If you'd implement my suggestion above, there is no longer a need for these tests here.
But I would argue that we'd need some checks in Verifier.cpp for these attributes, which would be more user-facing.

kmclaughlin-arm replied:
I've added some checks in the Verifier pass, for now these only check that incompatible attributes are not added to the same function.

kmclaughlin-arm committed:
- Added an enum (StateValue) to AArch64SMEAttributes for representing ZT0 attributes
- Added restrictions on the ZT0 attributes in the IR Verifier pass

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.h:
```diff
@@ -92,13 +97,22 @@ class SMEAttrs {
+  StateValue getZT0State() const {
+    return static_cast<StateValue>((Bitmask & ZT0_Mask) >> ZT0_Shift);
+  }
+  void setZT0State(StateValue S) {
+    Bitmask |= (static_cast<unsigned>(S) << ZT0_Shift);
+  }
```
To enable the use of the `SMEAttrs` constructor using e.g. `SMEAttrs(<mask>)`, perhaps it makes more sense to change `setZT0State` to something that encodes the value given a bitmask, i.e.

```
  static unsigned decodeZT0State(unsigned Bitmask) {
    return static_cast<StateValue>((Bitmask & ZT0_Mask) >> ZT0_Shift);
  }
  static unsigned encodeZT0State(StateValue S) {
    return static_cast<unsigned>(S) << ZT0_Shift;
  }
```

That avoids having to do:
```
   SA ZT0_In = SA(SA::Normal);
   ZT0_In.setZT0State(SA::StateValue::In);
```

and instead allows:
```
SA ZT0_In = SA(SA::encodeZT0State(SA::StateValue::In));
```
without the extra level of indirection.

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.h:
```diff
@@ -76,16 +86,35 @@ class SMEAttrs {
+  bool isInZT0() const { return getZT0State() == StateValue::In; }
+  bool isOutZT0() const { return getZT0State() == StateValue::Out; }
+  bool isInOutZT0() const { return getZT0State() == StateValue::InOut; }
+  bool preservesZT0() const { return getZT0State() == StateValue::Preserved; }
```
```suggestion
  bool isPreservesZT0() const { return getZT0State() == StateValue::Preserved; }
```

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -50,6 +54,22 @@ TEST(SMEAttributes, Constructors) {
                       ->getFunction("foo"))
                   .preservesZA());
 
+  ASSERT_TRUE(SA(*parseIR("declare void @foo() \"aarch64_sme_zt0_in\"")
```
Should we change the names of the attributes from `aarch64_sme_zt0_in` to the `aarch64_sme_in_zt0` format as well?  (and similar below)

kmclaughlin-arm replied:
I've changed the attributes to use this format, I think this makes sense given the Clang attribute will be `__arm_in("zt0")`

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.h:
```diff
@@ -76,16 +86,35 @@ class SMEAttrs {
+    Bitmask |= (static_cast<unsigned>(S) << ZT0_Shift);
+  }
+
+  bool hasNewZT0Body() const { return getZT0State() == StateValue::New; }
```
```suggestion
  bool isNewZT0() const { return getZT0State() == StateValue::New; }
```

sdesmalen-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2160,6 +2160,64 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
            V);
   }
 
+  if (Attrs.hasFnAttr("aarch64_sme_zt0_new")) {
```
Rather than having to spell out all possible combinations here, my preference would be to do the same thing as you did for the assert, where the error is more generic, but where the check itself is simpler by using the exclusive-or operator (and therefore less error prone to missing a case).

kmclaughlin-arm committed:
- Added decodeZT0State & encodeZT0State to enable the use of the SMEAttrs constructor with ZT0 attributes
- Changed preservesZT0() to isPreservesZT0()
- Changed hasNewZT0Body() to isNewZT0()
- Renamed attributes, for example aarch64_sme_zt0_new -> aarch64_sme_new_zt0
- Combined the warnings added to the IR Verifier

sdesmalen-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2160,6 +2160,21 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
+           Attrs.hasFnAttr("aarch64_sme_inout_zt0") ^
+           Attrs.hasFnAttr("aarch64_sme_out_zt0") ^
+           Attrs.hasFnAttr("aarch64_sme_preserved_zt0")),
+          "ZT0 state attributes 'aarch64_sme_[in|inout|out|new|preserved]_zt' "
```
nit: Just write out in full:
```
"Attributes 'aarch64_new_zt0', 'aarch64_in_zt0', 'aarch64_out_zt0', 'aarch64_inout_zt0' and 'aarch64_preserves_zt0' are mutually exclusive"
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.cpp:
```diff
@@ -18,16 +18,25 @@ void SMEAttrs::set(unsigned M, bool Enable) {
+  // ZT0 Attrs
+  assert((!sharesZT0() || (isNewZT0() ^ isInZT0() ^ isInOutZT0() ^ isOutZT0() ^
+                           isPreservesZT0())) &&
+         "ZT0_New, ZT0_In, ZT0_Out, ZT0_InOut and ZT0_Preserved are all "
+         "mutually exclusive");
```
```suggestion
         "Attributes 'aarch64_new_zt0', 'aarch64_in_zt0', 'aarch64_out_zt0', "
         "'aarch64_inout_zt0' and 'aarch64_preserves_zt0' are mutually exclusive");
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.cpp:
```diff
@@ -60,6 +69,16 @@ SMEAttrs::SMEAttrs(const AttributeList &Attrs) {
     Bitmask |= ZA_New;
   if (Attrs.hasFnAttr("aarch64_pstate_za_preserved"))
     Bitmask |= ZA_Preserved;
+  if (Attrs.hasFnAttr("aarch64_sme_in_zt0"))
```
Let's remove the `_sme` part from this, so that it's `aarch64_in_zt0`.

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.cpp:
```diff
@@ -60,6 +69,16 @@ SMEAttrs::SMEAttrs(const AttributeList &Attrs) {
+    Bitmask |= encodeZT0State(StateValue::Out);
+  if (Attrs.hasFnAttr("aarch64_sme_inout_zt0"))
+    Bitmask |= encodeZT0State(StateValue::InOut);
+  if (Attrs.hasFnAttr("aarch64_sme_preserved_zt0"))
```
nit: preserves (not preserved)

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -82,19 +102,86 @@ TEST(SMEAttributes, Basics) {
   ASSERT_TRUE(SA(SA::ZA_Shared | SA::ZA_Preserved).preservesZA());
 
   ASSERT_TRUE(SA(SA::ZA_New).hasPrivateZAInterface());
+  ASSERT_FALSE(SA(SA::ZA_New).hasSharedZAInterface());
```
nit: add this case too?
```
ASSERT_FALSE(SA(SA::ZA_Shared).sharesZA());
```

kmclaughlin-arm replied:
There is already a `sharesZA()` test for `SA::ZA_Shared` on line 105, did you mean:
`ASSERT_FALSE(SA(SA::ZA_Shared).sharesZT0())`?

I've added `sharesZT0()` cases for `SA::ZA_Shared` & `SA::ZA_New` in the latest commit.

sdesmalen-arm replied:
I'm not sure why I asked for this yesterday, perhaps my eyes just didn't spot this case.

kmclaughlin-arm committed:
- Simplified ZT0 attribute checks in Verifier.cpp
- Changed warning to use full names of all mutually exclusive attributes
- Removed 'sme' from attribute names, e.g. aarch64_sme_in_zt0 -> aarch64_in_zt0

kmclaughlin-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2160,6 +2160,21 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
+           Attrs.hasFnAttr("aarch64_sme_in_zt0") ^
+           Attrs.hasFnAttr("aarch64_sme_inout_zt0") ^
+           Attrs.hasFnAttr("aarch64_sme_out_zt0") ^
+           Attrs.hasFnAttr("aarch64_sme_preserved_zt0")),
```
This is much better, thanks :)
